include/MVC/SugarApplication.php

```diff
@@ -667,22 +667,23 @@ public static function headerRedirect($header_URL) {
     }
 
     /**
-     * Storing messages into session
+     * Add an error message for the user
      *
      * @access	public
-     * @param string $message
+     * @param string $error_message
      */
-    public static function appendErrorMessage($message) {
-        self::appendMessage('user_error_message', $message);
+    public static function appendErrorMessage($error_message)
+    {
+        SugarApplication::appendMessage($error_message, 'error');
```

Comment:
Do not change the session key as it already used as it is in many places.

Reply:
As these messages are not only "error messages" but also info, alert, etc.... I think we should stop using the session key "user_error_messages".
In this PR we use: `$_SESSION['suite_messages'][$type]`

I'm going to change this key everywhere:

```
/m/d/t/SuiteCRM git:develop_3048 ❯❯❯ git grep user_error_message                                                                                       
include/UploadFile.php:                if (!is_array($_SESSION['user_error_message'])) {
include/UploadFile.php:                    $_SESSION['user_error_message'] = array();
include/UploadFile.php:                $_SESSION['user_error_message'][] = $error_message;
include/UploadMultipleFiles.php:                if (!is_array($_SESSION['user_error_message'])) {
include/UploadMultipleFiles.php:                    $_SESSION['user_error_message'] = array();
include/UploadMultipleFiles.php:                $_SESSION['user_error_message'][] = $error_message;
modules/FP_events/controller.php:            $_SESSION['user_error_message'] = array();//clear the error message array
modules/FP_events/controller.php:            $_SESSION['user_error_message'] = array();
modules/FP_events/controller.php:            $_SESSION['user_error_message'] = array();
tests/unit/codecept/include/MVC/SugarApplicationTest.php:        //execute the method and check that the method adds the message to user_error_message array.
tests/unit/codecept/include/MVC/SugarApplicationTest.php:        $user_error_message_count = count($_SESSION['suite_messages']['error']);
tests/unit/codecept/include/MVC/SugarApplicationTest.php:        $this->assertGreaterThan($user_error_message_count, count($_SESSION['suite_messages']['error']
tests/unit/codecept/include/MVC/SugarApplicationTest.php:            $user_error_message_count = count($_SESSION['suite_messages'][$type]);
tests/unit/codecept/include/MVC/SugarApplicationTest.php:            $this->assertGreaterThan($user_error_message_count, count($_SESSION['suite_messages'][$typ

```

Reply:
You right, we should but we can't as it won't be backward compatible to any possibly exists customizations, so please don't do that.

Reply:
As I understand the only point that may have backwards incompatibility is where we read the $SESSION keys: user_error_message and user_success_message. And we do this in the method "getMessages()" in SugarApplication.... In develop branch this method is:

```php
    protected static function getMessages($type) {
        
        self::validateMessageType($type);

        if (isset($_SESSION[$type]) && is_array($_SESSION[$type])) {
            $msgs = $_SESSION[$type];
            unset($_SESSION[$type]);
            return $msgs;
        } else {
            return array();
        }
    }
```
And it is called like this in include/MVC/SugarApplication.php:

```php
$messages = self::getMessages('user_error_message');
$messages = self::getMessages('user_success_message');
```


And in my PR is:

```php
    public static function getMessages($type = null, $clear_queue = true)
    {
        $messages = array();
        if (isset($_SESSION['suite_messages'])) {
            if (isset($type)) {
                if (isset($_SESSION['suite_messages'][$type])) {
                    $messages[$type] = $_SESSION['suite_messages'][$type];
                    if ($clear_queue) {
                        unset($_SESSION['suite_messages'][$type]);
                    }
                }
            } else {
                $messages = $_SESSION['suite_messages'];
                if ($clear_queue) {
                    unset($_SESSION['suite_messages']);
                }
            }
        }
        }
        return($messages);
    }
    }
```

I will modify the method in my PR in order to do not break backwards compatibility for custom developments.

Reply:
Hi @gymad 
I agree with you, that's why I [commit some changes](https://github.com/salesagility/SuiteCRM/pull/6602/commits/d642489f9e4ec0d4c9cf13e0106b25026bd0ab3f) after your advice.

Don't you think this commit is working good in order to have backwards compatibility ?
Please let me know if you think is not good enough.

Reply:
Yes, maybe it could works but I still don't understand why you don't just following the original concept and adding some extra keys like `user_alert_message`, `user_info_message` and so on.. it makes your code complexity much more simple and you don't have to handle the "old message types"?

Reply:
When we developed this feature (many years ago), we didn't use the original concept because the messages are not all together in an array, there are as many arrays in $_SESSION as many error types.:

```php
$_SESSION['user_error_message']
$_SESSION['user_success_message']
etc
```
We took the idea from Drupal. As you can see in our original implementation in issue #3048 was a copy of [Drupal implementation](https://api.drupal.org/api/drupal/includes%21bootstrap.inc/function/drupal_set_message/7.x). After that, we improved the original implementation in closed PR #3078 and after almost 2 years sleeping, we re-write the PR in order to update it to the new codebase and to get the PR accepted. We also added @samus-aran [suggestion](https://github.com/salesagility/SuiteCRM/pull/3078#issuecomment-445176954)

In this PR the messages are all together in 'suite_messages' key.

`$_SESSION['suite_messages'][$type] = array();`

Managing the old messages types is not complex and also assures backwards compatibility:

https://github.com/salesagility/SuiteCRM/pull/6602/files#diff-3461d64b9728cb94323089f5674179f5R829

Reply:
@gymad I made a mistake and closed this conversation. I reopen it.